File: python/src/sort/shell_sort.py

def shell_sort(in_array):
    gap=len(in_array)
    while gap > 1:
        gap = get_gap(gap)
        i=gap
        while i < len(in_array):
            curr_value = in_array[i]
            j = i - gap
            while j > -1 and in_array[j] > curr_value:
                in_array[j+gap] = in_array[j]
                j -= gap
            in_array[j+gap] = curr_value
            i += gap 
    return in_array


def get_gap(gap):
    return int(gap * 10/11)

File: c#/Sort/ShellSort.cs
using System.Collections;

namespace Sort
{
    public class ShellSort : ISort
    {
        public ArrayList Sort(ArrayList inArray)
        {
            Gap = inArray.Count;
            while (Gap > 1)
            {
                Gap = Gap;
                int i = Gap;

                while(i < inArray.Count)
                {
                    int current = (int)inArray[i], j = i - Gap;
                    while(j>-1 && (int)inArray[j] > current)
                    {
                        inArray[j + Gap] = inArray[j];
                        j -= Gap;
                    }
                    inArray[j + Gap] = current;
                    i += Gap;
                }
            }
            return inArray;
        }

        private int _gap;
        public int Gap
        {
            get { return _gap; }
            set { _gap = value * 10 / 11; }
        }
    }
}
